mirgecom/steppers.py

```diff
@@ -80,11 +90,16 @@ def advance_state(rhs, timestepper, checkpoint, get_timestep,
         if dt < 0:
             return istep, t, state
 
-        checkpoint(state=state, step=istep, t=t, dt=dt)
+        if pre_step_callback is not None:
+            pre_step_callback(state=state, step=istep, t=t, dt=dt)
 
         state = timestepper(state=state, t=t, dt=dt, rhs=rhs)
 
         t += dt
+
+        if post_step_callback is not None:
+            post_step_callback(state=state, step=istep, t=t, dt=dt)
```

Comment:
I would advocate for having `post_step_callback` (and maybe `pre_step_callback` too?) return a boolean indicating whether to terminate timestepping. It gives the user some more flexibility in deciding when the simulation should stop. Then the timestepping loop can be a `while not done:` and the `t_final` stuff can be removed from the interface. I don't know if we want to actually make that change to the stepper in this PR, but in any case, just something to think about.

Reply:
I like this idea. How does a call signature like this look? `state, stop = callback(...)`, where `stop` is a bool?

Reply:
What about callbacks that raise exceptions though? That's my only real concern.

Reply:
I would advocate for placing the simulation termination in the hands of the user at the callback level.  Consider this callback:

```python
def my_callback(state, step, t, dt):
   is_healthy = my_healthcheck(state, step, t, dt)
   is_healthy = comm.allreduce(is_healthy, MPI.LOR)
   if not is_healthy:
      write_viz(...)
      write_restart(...)
      do_my_other_biddings(...)
      raise SimError("Simulation health check failed.")
```

This is safe because the callback routine is already collective. If we are concerned with users doing collective operations (I don't see why we should be) - then we can add a wrapper.  I like this a lot because it places healthcheck at the driver level, places full control over how to handle a health issue in the hands of the user (correctly imo), and forces users to opt-in to use it.

Reply:
> I would advocate for placing the simulation termination in the hands of the user at the callback level. Consider this callback:
> 
> ```python
> def my_callback(state, step, t, dt):
>    is_healthy = my_healthcheck(state, step, t, dt)
>    is_healthy = comm.allreduce(is_healthy, MPI.LOR)
>    if not is_healthy:
>       write_viz(...)
>       write_restart(...)
>       do_my_other_biddings(...)
>       raise SimError("Simulation health check failed.")
> ```
> 
> This is safe because the callback routine is already collective. If we are concerned with users doing collective operations (I don't see why we should be) - then we can add a wrapper. I like this a lot because it places healthcheck at the driver level, places full control over how to handle a health issue in the hands of the user (correctly imo), and forces users to opt-in to use it.

I think I mostly agree with this. My only complaint is that when doing it this way the user won't get any information back from `my_healthcheck` about what exactly failed. My inclination would be to make `my_healthcheck` collective and then either:

1. Print out what failed on rank 0 inside `my_healthcheck`
2. Do something like:

```python
def my_callback(state, step, t, dt):
   try:
      my_healthcheck(state, step, t, dt)
   except HealthCheckError:
      write_viz(...)
      write_restart(...)
      do_my_other_biddings(...)
      raise
```
(i.e., whatever is contained in the HealthCheckError will bubble back up to the user.)

Reply:
> My only complaint is that when doing it this way the user won't get any information back from `my_healthcheck` about what exactly failed.

Agreed. If we have a healthcheck utility, say in `simutils`, then it could potentially return a string about what went wrong. `is_healthy, status_message = my_healthcheck(...)`.  Just another alternative.  I like the idea of handing that string to the user and letting them figure out what to do with it at the driver level... but that's just me.

Reply:
If we catch the exception using what @majosm proposed (try - except clause), then we can yank out the error message from the exception and print that out on rank 0. How does that sound?

Reply:
> What about callbacks that raise exceptions though? That's my only real concern.

I think exceptions from callbacks are OK if the user handles it right.  Callbacks are collective already.  Clearly, the user can toss an exception from only one rank - but hey: *that's on the user*.

Reply:
> If we catch the exception using what @majosm proposed (try - except clause), then we can yank out the error message from the exception and print that out on rank 0. How does that sound?

I think that is OK, but imo not as powerful as forming a status string and letting the user deal with the error.  It means that the healthcheck does more than just check the health - it also does I/O on rank 0 - which may or may not be what the user wants to do there.

Reply:
> > If we catch the exception using what @majosm proposed (try - except clause), then we can yank out the error message from the exception and print that out on rank 0. How does that sound?
> 
> I think that is OK, but imo not as powerful as forming a status string and letting the user deal with the error. It means that the healthcheck does more than just check the health - it also does I/O on rank 0 - which may or may not be what the user wants to do there.

Sorry, I meant doing one or the other, i.e., _either_ have `my_healthcheck` print, _or_ do the `try`/`except` thing (my preference is the latter). With the latter approach you could extract the error message by doing something like:

```python
try:
    my_healthcheck(state, step, t, dt)
except HealthCheckError as e:
    msg = e.message
    # ... do more processing with msg
```

Reply:
> Sorry, I meant doing one or the other, i.e., _either_ have `my_healthcheck` print, _or_ do the `try`/`except` thing (my preference is the latter). With the latter approach you could extract the error message by doing something like:
> 
> ```python
> try:
>     my_healthcheck(state, step, t, dt)
> except HealthCheckError as e:
>     msg = e.message
>     # ... do more processing with msg
> ```

I like this OK. What I would like to avoid is needing to pass "comm" as an argument to the healthcheck utility.  So if `my_healthcheck` essentially calls a library routine underneath but still does synchronization (i.e. collective error handling) at driver level, then I think that is better than doing collective health check and collective exception in the lower layer (i.e. in the library or simutils).